### PR TITLE
sc-consensus-manual-seal uses Finalizer trait for finalization instead of Backend.

### PR DESCRIPTION
**Pull request by seunlanlege**
Using the `Backend` trait only allows for finalization one block at a time. The `Finalizer` trait allows for finalizing a subtree.

### CODE REVIEW

File: client/consensus/manual-seal/src/lib.rs [hunk 1]
```diff
@@ -28,55 +29,28 @@ use sp_consensus::{
 		BoxBlockImport,
 	},
 };
+use sp_blockchain::{HeaderBackend};
```
Comment (JoshOrndorff) [inline client/consensus/manual-seal/src/lib.rs:32]:
```suggestion
use sp_blockchain::HeaderBackend;
```

**Comment by JoshOrndorff** [comment]
> It would be really helpful if you tried to integrate this into either node or node-template (just on your own local branch)

That is what https://github.com/substrate-developer-hub/recipes/pull/202 strives to address. When it is done it will be a complete working example of how to use instant and manual seal with an accompanying writeup.

File: client/consensus/manual-seal/src/finalize_block.rs [hunk 2]
```diff
@@ -23,35 +23,40 @@ use sp_runtime::{
...
 	pub hash: <B as BlockT>::Hash,
 	/// sender to report errors/success to the rpc.
 	pub sender: rpc::Sender<()>,
 	/// finalization justification
 	pub justification: Option<Justification>,
 	/// client backend
-	pub backend: Arc<CB>,
+	pub finalizer: Arc<F>,
+	/// phantom type
+	pub _phantom: PhantomData<CB>,
```
[the remark this answers is not in the document]
Comment (seunlanlege) [reply client/consensus/manual-seal/src/finalize_block.rs:40]: The `Finalizer` is generic over the `Backend` trait 

https://github.com/paritytech/substrate/blob/fca34f6180522724e318f14a32980102e6199958/client/api/src/backend.rs#L184

File: client/consensus/manual-seal/src/lib.rs [hunk 3]
```diff
@@ -28,55 +29,28 @@ use sp_consensus::{
...
+};
+pub use self::{
+	error::Error,
+	rpc::{EngineCommand, CreatedBlock},
+};
+use sc_client_api::{TransactionFor, StateBackend, Backend};
+use sp_runtime::traits::HashFor;
 
 /// The verifier for the manual seal engine; instantly finalizes.
 struct ManualSealVerifier;
```
[the remark this answers is not in the document]
Comment (seunlanlege) [reply client/consensus/manual-seal/src/lib.rs:56]: We construct the `BasicQueue` with this verifier.
> This won't work anyway if we create an import queue with ManualSealVerifier but with BabeBlockImport.

I'm not really sure what this means or why we would do this. :sweat_smile:

**Comment by seunlanlege** [comment]
@andresilva hey, could you please take a look at this again?

File: client/consensus/manual-seal/src/finalize_block.rs [hunk 4]
```diff
@@ -23,35 +23,40 @@ use sp_runtime::{
...
 /// params for block finalization.
-pub struct FinalizeBlockParams<B: BlockT, CB> {
+pub struct FinalizeBlockParams<B: BlockT, F, CB> {
 	/// hash of the block
 	pub hash: <B as BlockT>::Hash,
 	/// sender to report errors/success to the rpc.
 	pub sender: rpc::Sender<()>,
 	/// finalization justification
 	pub justification: Option<Justification>,
 	/// client backend
```
Comment (bkchr) [inline client/consensus/manual-seal/src/finalize_block.rs:37]: Docs outdated.

**Comment by andresilva** [comment]
check polkadot failed due to unrelated reason.

**Comment by bkchr** [comment]
@seunlanlege please merge master to make the polkadot check happy.

File: client/consensus/manual-seal/src/seal_new_block.rs [hunk 5]
```diff
@@ -54,37 +53,37 @@ pub struct SealBlockParams<'a, B: BlockT, C, CB, E, T, P: txpool::ChainApi> {
 	pub sender: rpc::Sender<CreatedBlock<<B as BlockT>::Hash>>,
 	/// transaction pool
 	pub pool: Arc<txpool::Pool<P>>,
-	/// client backend
-	pub backend: Arc<CB>,
+	/// HeaderBackend trait
```
Comment (andresilva) [inline client/consensus/manual-seal/src/seal_new_block.rs:56]:
```suggestion
	/// header backend
```

**Comment by tarunjais2** [comment]
hi @bkchr, @JoshOrndorff can we integrate gossips protocols for finalisation of blocks in instant-seal consensus? just like it is working for grandpa consensus?

**Comment by bkchr** [comment]
What gossip protocols? What you want to do?

**Comment by JoshOrndorff** [comment]
You can exactly use grandpa for finality and instant seal for authoring. They work together.

**Comment by tarunjais2** [comment]
> What gossip protocols? What you want to do?

I want to author blocks by instant seal so that the blocks will be created only when some transactions happened and it will finalize only after grandpa gossip

**Comment by tarunjais2** [comment]
> You can exactly use grandpa for finality and instant seal for authoring. They work together.

okay I will look into it

**Comment by tarunjais2** [comment]
thanks @JoshOrndorff it is working

**Pull request merged**